src/background_thread.c

```diff
@@ -661,36 +667,44 @@ background_thread_prefork1(tsdn_t *tsdn) {
 	}
 }
 
-static void
-background_thread_postfork_init(tsdn_t *tsdn) {
-	assert(n_background_threads == 0);
-	if (background_thread_enabled_at_fork) {
-		background_thread_enabled_set(tsdn, true);
-		background_threads_enable(tsdn_tsd(tsdn));
-	}
-}
-
 void
 background_thread_postfork_parent(tsdn_t *tsdn) {
 	for (unsigned i = 0; i < ncpus; i++) {
 		malloc_mutex_postfork_parent(tsdn,
 		    &background_thread_info[i].mtx);
 	}
-	background_thread_postfork_init(tsdn);
 	malloc_mutex_postfork_parent(tsdn, &background_thread_lock);
 }
 
+static void
+background_thread_postfork_init(tsdn_t *tsdn) {
+	if (!background_thread_enabled_at_fork) {
+		return;
+	}
+	malloc_mutex_lock(tsdn, &background_thread_lock);
+	/* Clear background_thread state (it's disabled after fork). */
+	n_background_threads = 0;
+	background_thread_enabled_set(tsdn, false);
+	for (unsigned i = 0; i < ncpus; i++) {
+		background_thread_info_t *info = &background_thread_info[i];
+		malloc_mutex_lock(tsdn, &info->mtx);
+		info->started = false;
+		pthread_cond_destroy(&info->cond);
```

Comment:
The hang was actually because of this. We only need to take care of the cond vars in the child, and that is pretty much to re-initialize them, as we don't intend to re-enable the threads. So the destroy wasn't really safe. We should just pthread_cond_init as it's irrelevant in the child process.